torch_geometric/contrib/nn/models/graph_lottery_ticket.py

```diff
@@ -0,0 +1,515 @@
+
+    GLTSearch class provides an interface to prune a model to a specified
+    graph/model pruning rate.
+
+    This methodology is built for both node classification and link
+    prediction, but any other tasks should be easily extensible given the
+    appropriate loss function and data.
+
+    .. note::
+        For examples of using the GLTSearch, see
+        `examples/contrib/graph_lottery_ticket.py
+        <https://github.com/pyg-team/pytorch_geometric/blob/master/
+        examples/contrib/graph_lottery_ticket.py>`_.
+
+    Args:
+        module (torch.nn.Module): The GNN module to prune.
+        graph (torch_geometric.data.Data): Graph to perform GLT search over.
+        lr (float): Learning rate for training. Is propagated to masks if
+        individual lr not specified.
+        reg_graph (float): L2 regularization for graph mask.
+        reg_model (float): L2 regularization for model masks.
+        optim_args (dict): Args for internal optimizer.
+        task (str): Specifies to use node or link train loop.
+        lr_mask_model (float, optional):  LR to apply to model mask only.
+        lr_mask_graph Optional[float] LR to apply to graph mask only.
+        optimizer (torch.optim.Optimizer): Optimizer to use for training.
+        prune_rate_model (float): Sparsity level induced in model masks.
+        prune_rate_graph (float): Sparsity level induced in graph mask.
+        max_train_epochs (int): max number of epochs to train.
+        loss_fn (callable): loss function to train with.
+        save_all_masks (bool): toggles saving all masks.
+        seed (int): random seed.
+        verbose (bool): toggles trainer verbosity.
+        ignore_keys (set, optional): Set of keys to ignore when injecting
+        masks into the model.
+    """
+    module: Module
+    graph: Data
+    lr: float
+    reg_graph: float
+    reg_model: float
+    task: str
+    optim_args: Dict[str, Any] = field(default_factory=lambda: {})
+    lr_mask_model: Optional[float] = None
+    lr_mask_graph: Optional[float] = None
+    optimizer: Type[Optimizer] = Adam
+    prune_rate_model: float = 0.2
+    prune_rate_graph: float = 0.05
+    max_train_epochs: int = 200
+    loss_fn: Callable = cross_entropy
+    save_all_masks: bool = False
+    seed: int = field(default_factory=lambda: randint(1, 9999))
+    verbose: bool = False
+    ignore_keys: set = field(default_factory=lambda: set())
+
+    def __post_init__(self):
+        """fixes internal hyperparams and generates GLTMask from
+        :obj:`GLTMask` from model """
+        torch.manual_seed(self.seed)
+
+        if not self.lr_mask_graph:
+            self.lr_mask_graph = self.lr
+
+        if not self.lr_mask_model:
+            self.lr_mask_model = self.lr
+
+        self.optim_args = {"lr": self.lr, **self.optim_args}
+
+        self.mask = GLTMask(
+            self.module, self.graph,
+            torch.device("cuda" if torch.cuda.is_available() else "cpu"))
```

Comment:
Why is this on the GPU. It should be on the same device as `self.module`.

Reply:
Addressed!